fenzo-core/src/main/java/com/netflix/fenzo/ShortfallEvaluator.java

```diff
@@ -19,113 +19,15 @@
-        // maintain the list of tasks for which we have already requested resources. So, we ask again only if we notice
-        // new tasks. There can be some scenarios where a resource that we asked got used up for a task other than the one
-        // we asked for. Although, this is unlikely in FIFO queuing of tasks. So, we maintain the list of tasks with a
-        // timeout and unqueueTask it from there so we catch this scenario and ask for the resources again for that task.
-
-        final HashMap<String, Integer> shortfallMap = new HashMap<>();
-        if (attrKeys != null && failures != null && !failures.isEmpty()) {
-            removeOldInserts();
-            long now = System.currentTimeMillis();
-            for (TaskRequest r : failures) {
-                String tid = r.getId();
-                if (requestedForTasksSet.get(tid) == null) {
-                    requestedForTasksSet.put(tid, now);
-                    fillShortfallMap(shortfallMap, attrKeys, r);
-                }
-            }
-        }
-
-        return adjustAgentScaleUp(shortfallMap, autoScaleRules);
-    }
-
-    private Map<String, Integer> adjustAgentScaleUp(HashMap<String, Integer> shortfallMap, AutoScaleRules autoScaleRules) {
-        Map<String, Integer> corrected = new HashMap<>(shortfallMap);
-        for (Map.Entry<String, Integer> entry : shortfallMap.entrySet()) {
-            AutoScaleRule rule = autoScaleRules.get(entry.getKey());
-            if (rule != null) {
-                int adjustedValue = rule.getShortfallAdjustedAgents(entry.getValue());
-                if (adjustedValue > 0) {
-                    corrected.put(entry.getKey(), adjustedValue);
-                }
-            }
-        }
-        return corrected;
-    }
+/* package */ interface ShortfallEvaluator {
 
-    private void fillShortfallMap(HashMap<String, Integer> shortfallMap, Set<String> attrKeys, TaskRequest r) {
-        for (String k : attrKeys) {
-            if (matchesTask(r, k)) {
-                if (shortfallMap.get(k) == null)
-                    shortfallMap.put(k, 1);
-                else
-                    shortfallMap.put(k, shortfallMap.get(k) + 1);
-            }
-        }
-    }
+    void setTaskToClustersGetter(Func1<QueuableTask, List<String>> getter);
 
-    private boolean matchesTask(TaskRequest r, String k) {
-        if (!(r instanceof QueuableTask) || taskToClustersGetter == null)
-            return true;
-        final List<String> strings = taskToClustersGetter.call((QueuableTask) r);
-        if (strings != null && !strings.isEmpty()) {
-            for (String s : strings)
-                if (k.equals(s))
-                    return true;
-            return false; // doesn't match
-        }
-        // matches anything
-        return true;
-    }
+    Map<String, Integer> getShortfall(Set<String> attrKeys, Set<TaskRequest> failures, AutoScaleRules autoScaleRules);
```

Comment:
attrKeys name is very cryptic. I have figured out that it contains something like set of cluster names. If you split this name there are two parts:
* 'attr' - is an intent here to pass any host/cluster attribute value?
* 'Keys' - these are not really keys, but values of attributes
If we want to keep this name, we should provide better explanation of the actual values that can be put there, and what is their source.